extension/source/Controllers/Account/AccountTrackerController.ts

```diff
@@ -0,0 +1,155 @@
+
+  private blockTracker: PollingBlockTracker;
+
+  private mutex = new Mutex();
+
+  private ethQuery: EthQuery;
+
+  private getIdentities: () => PreferencesState['identities'];
+
+  private getCurrentChainId: NetworkController['getNetworkIdentifier'];
+
+  constructor({
+    config,
+    state,
+    provider,
+    blockTracker,
+    getCurrentChainId,
+    getIdentities,
+    onPreferencesStateChange,
+  }: {
+    config: AccountTrackerConfig;
+    state: Partial<AccountTrackerState>;
+    provider: SafeEventEmitterProvider;
+    blockTracker: PollingBlockTracker;
+    getCurrentChainId: NetworkController['getNetworkIdentifier'];
+    getIdentities: () => PreferencesState['identities'];
+    onPreferencesStateChange: (
+      listener: (preferencesState: PreferencesState) => void,
+    ) => void;
+  }) {
+    super({ config, state });
+    this.defaultState = {
+      accounts: {},
+    };
+    this.defaultConfig = {
+      _currentBlock: '',
+    };
+    this.initialize();
+    this.provider = provider;
+    this.blockTracker = blockTracker;
+    this.ethQuery = new EthQuery(provider);
+
+    // This starts the blockTracker internal tracking
+    this.blockTracker.on('latest', (block: string) => {
+      this.configure({ _currentBlock: block });
+      this.refresh();
+    });
+
+    this.getIdentities = getIdentities;
+    this.getCurrentChainId = getCurrentChainId;
+    onPreferencesStateChange(() => {
+      this.syncAccounts();
+      this.refresh();
+    });
+  }
+
+  syncAccounts(): void {
+    const { accounts } = this.state;
+    const addresses = Object.keys(this.getIdentities());
+    const existing = Object.keys(accounts);
+    const newAddresses = addresses.filter(
+      (address) => existing.indexOf(address) === -1,
+    );
+    const oldAddresses = existing.filter(
+      (address) => addresses.indexOf(address) === -1,
+    );
+    newAddresses.forEach((address) => {
+      accounts[address] = { balance: '0x0' };
+    });
+    oldAddresses.forEach((address) => {
+      delete accounts[address];
+    });
+    this.update({ accounts: { ...accounts } });
+  }
+
+  async refresh(): Promise<void> {
+    const releaseLock = await this.mutex.acquire();
+    try {
+      const { accounts } = this.state;
+      const currentBlock = this.config._currentBlock;
+      if (!currentBlock) return;
+      const addresses = Object.keys(accounts);
+      if (addresses.length > 0) {
+        await Promise.all(
+          addresses.map((x) => this._updateAccount(x, currentBlock)),
+        );
+      }
+    } catch (error) {
+      // ignore error
+    } finally {
+      releaseLock();
+    }
+  }
+
+  private async _updateAccount(
+    address: string,
+    currentBlock: string,
+  ): Promise<void> {
+    const currentChainId = this.getCurrentChainId();
+    if (currentChainId === 'loading') {
+      return;
+    }
+    return new Promise((resolve, reject) => {
+      // query balance
+      this.ethQuery.sendAsync(
```

Comment:
What's the idea behind `ethQuery`? Can't we use ethers directly?
https://docs.ethers.io/v5/api/providers/provider/#Provider-getBalance

Reply:
It's by design. ethers is a bit heavy handed in it's approach to blockchain calls.
we use ethQuery within controllers for 3 reasons:
1 - light weight processing (No listeners added to provider)
2- goes through our rpc engine. (can do additional processing inside our middlewares)
3 - uses fetch middleware at the end to communicate with rpc endpoint (which has retry logic built-in)